src/main/scala/asciibot/parser/parser.scala

```diff
@@ -0,0 +1,200 @@
+
+    val surrs = str_surrs map (_ match {
+        case (_, "#") => Blocked
+        case (_, "_") => Open
+        case (_, "*") => Anything
+        case (dir, s) => throw new CharBlockError(s"Invalid Surrounding `$s' for $dir side", start, end)
+      })
+
+    (Surroundings(surrs(0), surrs(1), surrs(2), surrs(3)), State(state_str))
+  }
+
+  def parseOutState(a: Seq[String],
+                    start: Int,
+                    end: Int,
+                    surrs: Surroundings) : (MoveDirection, State) = {
+    def throwBE(m : String) = throw new CharBlockError(m, start, end)
+  
+    //println(s"parseOutState: $a")
+    assert(a.length == 3)
+
+    val sym = "[#_*]"
+    val id = "[^#_*/\\-> |]*"
+    // >0 spaces, then either an #, o, or *, or a state name, then >0 spaces
+    val topBotRegex = s"""^ +(${id}|${sym}) +$$""".r
+    // id or sym, id or spaces, id or sym
+    val midRegex = s"""^(${id}|${sym})(${id}| +)(${id}|${sym})$$""".r
+
+    val tmo = topBotRegex findPrefixMatchOf a(0).slice(start,end)
+    val mmo = midRegex findPrefixMatchOf a(1).slice(start,end)
+    val bmo = topBotRegex findPrefixMatchOf a(2).slice(start,end)
+
+    // Unwrap options and make sure the regex was matched
+    val tm = tmo getOrElse throwBE("Top row malformed")
+    val mm = mmo getOrElse throwBE("Middle row malformed")
+    val bm = bmo getOrElse throwBE("Bottom row malformed")
+
+    // Verify column alignment
+    if (tm.start(1) < mm.start(2) || tm.end(1) > mm.end(2)) {
+      throwBE("top and middle not aligned")
+    }
+    if (bm.start(1) < mm.start(2) || bm.end(1) > mm.end(2)) {
+      throwBE("middle and bottom not aligned")
+    }
+    
+    // Strings that are in the north, east, west, south, and center positions
+    val matches = List(tm.group(1), mm.group(3), mm.group(1), bm.group(1), mm.group(2))
+
+    println("matches are " +  matches)
+
+    // Make sure only one state is specified
+    val numOutputStates = matches count {s => !("#_* " contains s(0))}
+    if ( 1 != numOutputStates ) {
+      throwBE(s"$numOutputStates output states were provided")
+    }
+
+    // Make sure that output surroundings match input surroundings
+    val Surroundings(inTop, inRight, inLeft, inBot) = surrs
+    val inSurrList = List(inTop, inRight, inLeft, inBot)
+    inSurrList zip matches.slice(0,4) foreach { _ match {
+      case (Blocked, "#") => ()
+      case (Open, "_") => ()
+      case (Anything, "*") => ()
+      case (_, s) if !("#_*" contains s(0)) => () // Surrounding in input, state in output
+      case (i,o) => throwBE(s"Surrounding mismatch: ${surrToChar(i)} to $o")
+      }
+    }
+
+    val moveDirIdx = matches indexWhere (s => !("#_*" contains s(0)))
+    val moveDir = List(North, East, West, South, StayHere)(moveDirIdx)
+    val nextState = State(matches(moveDirIdx))
+    (moveDir, nextState)
+  }
+
+  def surrToChar(s: RelativeDescription) : String = s match {
+    case Blocked => "#"
+    case Open => "_"
+    case Anything => "*"
```

Comment:
Very good remark Zoab. Ideally, you could have separated the parser from the interpreter, but it's neat that you implemented all this.